cpp/include/cudf/detail/merge.hpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2018-2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/types.hpp>
+#include <cudf/table/row_operators.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail{
+
+
+/**
+ * @brief Source table identifier to copy data from.
+ */
+enum class side : bool { LEFT, RIGHT };
+
+/**
+ * @brief Tagged index type: `thrust::get<0>` indicates left/right side, 
+ * `thrust::get<1>` indicates the row index
+ */  
+using index_type = thrust::tuple<side, cudf::size_type>;  
+
+/**
+ * @brief tagged_element_relational_comparator uses element_relational_comparator to provide "tagged-index" comparation logic.
+ *
+ * Special treatment is necessary in several thrust algorithms (e.g., merge()) where 
+ * the index affinity to the side is not guaranteed; i.e., the algorithms rely on 
+ * binary functors (predicates) where the operands may transparently switch sides.
+ *
+ * For example, 
+ *         thrust::merge(left_container,
+ *                       right_container,
+ *                       predicate(lhs, rhs){...});
+ *         can create 4 different use-cases, inside predicate(...):
+ *
+ *         1. lhs refers to the left container; rhs to the right container;
+ *         2. vice-versa;
+ *         3. both lhs and rhs actually refer to the left container;
+ *         4. both lhs and rhs actually refer to the right container;
+ *
+ * Because of that, one cannot rely on the predicate having *fixed* references to the containers.
+ * Each invocation may land in a different situation (among the 4 above) than any other invocation.
+ * Also, one cannot just manipulate lhs, rhs (indices) alone; because, if predicate always applies
+ * one index to one container and the other index to the other container, 
+ * switching the indices alone won't suffice in the cases (3) or (4), 
+ * where the also the containers must be changed (to just one instead of two) 
+ * independently of indices;
+ *
+ * As a result, a special comparison logic is necessary whereby the index is "tagged" with side information
+ * and consequently comparator functors (predicates) must operate 
+ * on these tagged indices rather than on raw indices.
+ *  
+ */
+template <bool has_nulls = true>
+struct tagged_element_relational_comparator {
+
+  __host__ __device__
```

Comment:
Should this file be renamed to `merge.cuh` to make it more clear that it contains device code?

Reply:
There's already a `merge.cuh` in strings (to be tackled next). Also, there was a specific request for `merge.hpp` name in a previous comment review.

Reply:
May I resolve this?

Reply:
@jrhemstad @harrism Can you clarify what the convention should be for naming header files that contain device code?

Reply:
What was the previous comment requesting this to be .hpp?

Reply:
"This should not be part of the public merge.hpp file. It should be in the detail/merge.hpp." (Jake) 
See email for the actual request.

Reply:
Not being part of the public header is fine, why not detail/merge.cuh? Jake said that headers containing device code should be .cuh.

Reply:
Okay, I'll do that, then.

Reply:
Done.